autonat.go

```diff
@@ -0,0 +1,147 @@
+package autonat
+
+import (
+	"context"
+	"errors"
+	"math/rand"
+	"sync"
+	"time"
+
+	host "github.com/libp2p/go-libp2p-host"
+	peer "github.com/libp2p/go-libp2p-peer"
+	ma "github.com/multiformats/go-multiaddr"
+)
+
+type NATStatus int
+
+const (
+	NATStatusUnknown NATStatus = iota
+	NATStatusPublic
+	NATStatusPrivate
+)
+
+var (
+	AutoNATBootDelay       = 15 * time.Second
+	AutoNATRefreshInterval = 15 * time.Minute
+)
+
+type AutoNAT interface {
+	Status() NATStatus
+	PublicAddr() (ma.Multiaddr, error)
+}
+
+type AmbientAutoNAT struct {
+	ctx    context.Context
+	host   host.Host
+	peers  map[peer.ID]struct{}
+	status NATStatus
+	addr   ma.Multiaddr
+	mx     sync.Mutex
+}
+
+func NewAutoNAT(ctx context.Context, h host.Host) AutoNAT {
+	as := &AmbientAutoNAT{
+		ctx:    ctx,
+		host:   h,
+		peers:  make(map[peer.ID]struct{}),
+		status: NATStatusUnknown,
+	}
+
+	h.Network().Notify(as)
+	go as.background()
+
+	return as
+}
+
+func (as *AmbientAutoNAT) Status() NATStatus {
+	return as.status
+}
+
+func (as *AmbientAutoNAT) PublicAddr() (ma.Multiaddr, error) {
+	as.mx.Lock()
+	defer as.mx.Unlock()
+
+	if as.status != NATStatusPublic {
+		return nil, errors.New("NAT Status is not public")
+	}
+
+	return as.addr, nil
+}
+
+func (as *AmbientAutoNAT) background() {
+	// wait a bit for the node to come online and establish some connections
+	// before starting autodetection
+	time.Sleep(AutoNATBootDelay)
+	for {
+		as.autodetect()
+		select {
+		case <-time.After(AutoNATRefreshInterval):
+		case <-as.ctx.Done():
+			return
+		}
+	}
+}
+
+func (as *AmbientAutoNAT) autodetect() {
+	if len(as.peers) == 0 {
+		log.Debugf("skipping NAT auto detection; no autonat peers")
+		return
+	}
+
+	as.mx.Lock()
+	peers := make([]peer.ID, 0, len(as.peers))
+	for p := range as.peers {
+		if len(as.host.Network().ConnsToPeer(p)) > 0 {
+			peers = append(peers, p)
+		}
+	}
+
+	if len(peers) == 0 {
+		// we don't have any open connections, try any autonat peer that we know about
+		for p := range as.peers {
+			peers = append(peers, p)
+		}
+	}
+
+	as.mx.Unlock()
+
+	shufflePeers(peers)
+
+	for _, p := range peers {
+		cli := NewAutoNATClient(as.host, p)
+		ctx, cancel := context.WithTimeout(as.ctx, 60*time.Second)
```

Comment:
magic number

Reply:
we can give a name to the incantation.

Reply:
named it and made it a variable so that we can unit test; https://github.com/libp2p/go-libp2p-autonat/pull/1/commits/fa14117521021ef03db7503a709e26a0f34c49a6